crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+                        hashable_params(&second_instr.params),
+                    ),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+            })
+            .or_insert_with(|| {
+                let mut entries = HashMap::with_capacity(1);
+                let key = (
+                    Self::get_relative_placement(first_qargs, second_qargs),
+                    (
+                        hashable_params(&first_instr.params),
+                        hashable_params(&second_instr.params),
+                    ),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+                CommutationCacheEntry { mapping: entries }
+            });
+        is_commuting
+    }
+
+    fn commute_matmul(
+        &self,
+        first_instr: &CircuitInstruction,
+        first_qargs: &[usize],
+        second_instr: &CircuitInstruction,
+        second_qargs: &[usize],
+    ) -> bool {
+        // compute relative positioning of qargs of the second gate to the first gate
+        let mut qarg: HashMap<&usize, usize> =
+            HashMap::with_capacity(first_qargs.len() + second_qargs.len());
+        for (i, q) in first_qargs.iter().enumerate() {
+            qarg.entry(q).or_insert(i);
+        }
+        let mut num_qubits = first_qargs.len();
+        for q in second_qargs {
+            if !qarg.contains_key(q) {
+                qarg.insert(q, num_qubits);
+                num_qubits += 1;
+            }
+        }
+
+        let first_qarg: Vec<_> = first_qargs
+            .iter()
+            .map(|q| *qarg.get_item(q).unwrap())
+            .collect();
+        let second_qarg: Vec<_> = second_qargs
+            .iter()
+            .map(|q| *qarg.get_item(q).unwrap())
+            .collect();
+
+        assert!(
+            first_qarg.len() <= second_qarg.len(),
+            "first instructions must have at most as many qubits as the second instruction"
+        );
+
+        let first_op = first_instr.op();
+        let second_op = second_instr.op();
+
+        let first_mat = match first_op.matrix(&first_instr.params) {
+            Some(mat) => mat,
+            None => match first_op {
+                PyGateType(gate) => match Python::with_gil(|py| -> Option<Array2<Complex64>> {
+                    Self::get_op(py, &gate.gate)
+                }) {
+                    Some(x) => x,
+                    _ => return false,
+                },
+                PyOperationType(operation) => {
+                    match Python::with_gil(|py| -> Option<Array2<Complex64>> {
+                        Self::get_op(py, &operation.operation)
+                    }) {
+                        Some(x) => x,
+                        _ => return false,
+                    }
+                }
+                _ => return false,
+            },
```

Comment:
Try not to use `Python::with_gil()` as it involves acquiring the gil manually from the `Rust` side. If you absolutely need to use python in a function, make sure you pass `py: Python` as one of the function arguments, in this case:

```py
fn commute_matmul(
        &self,
        py: Python,
        first_instr: &CircuitInstruction,
        first_qargs: &[usize],
        second_instr: &CircuitInstruction,
        second_qargs: &[usize],
    ) -> bool {
```

Reply:
Yeah I was worried about this, too. However, we would hopefully not need the gil too often; we'd only invoke it if there is a `Pygate` or `PyOperation` that does not define a `to_matrix`. Is there a performance benefit of acquiring the gil manually from Rust side if it does not happen too often vs setting `Python` as a parameter to the function? I had expected this is a 'we sometime acquire gil'- vs 'we always acquire gil'-type-of-scenario.